README.md
Austrian Lobbyregister Scraper
=============================================================

The scraper extracts data from the [Austrian Lobbying Register](http://www.lobbyreg.justiz.gv.at), which must be reported since 2013. The data are stored in CSV and JSON files to make the further usage as easy as possible.

This repository provides the code and documentation and [keeps track of bugs as well as feature requests](https://github.com/OKFNat/lobbyScraper/issues).

- [Data Source](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/liste!OpenForm&subf=a)
- Team: [Gute Taten für gute Daten](http://okfn.at/gutedaten/) Project (Open Knowledge Austria)
- Status: Production
- Documentation: English
- License:
	- Content: [Creative Commons Attribution 4.0](http://creativecommons.org/licenses/by/4.0/)
	- Software: [MIT License](http://opensource.org/licenses/MIT) 

**Used software**

The sourcecode is written in Python 2. It was created with use of [iPython](http://ipython.org/), [BeautifulSoup4](http://www.crummy.com/software/BeautifulSoup/) and [urllib2](https://docs.python.org/2/library/urllib2.html).

## SCRAPER

**Description**

The scraper fetches the overview page with the table and parses out the data with beautifulsoup4. Then the scraper goes into every lobbying register entry via the Registerzahl-link, stores the html and parses out the rest of the data available via regular expressions. At the end, the data is stored as JSON and CSV files for easy usage later on.

**Run scraper**

Go into the root folder of this repository and execute following commands in your terminal:
```
cd code
python lobbyscraper.py
```

### How the scraper works

**Configure the Scraper**

There are two global variables in [lobbyscraper.py](code/lobbyscraper.py) you may want to change to your needs.

- DELAY_TIME: To not overload the server or may get blocked because of too many request, you should set the delay time to fetch to 1-5 seconds, not less.
- TS: The timestamp as a string can be set to the last download. So you can use downloaded data over and over again and must not do it everytime. When you do it first time, you can set the value to ```datetime.now().strftime('%Y-%m-%d-%H-%M')```, so it is the timestamp when the scraper starts.

**Download raw html**

Here all the html raw data gets downloaded, stored locally and the basic data gets parsed.

- Download the overview page with the tables (html).
- Open the downloaded file.
- Parse out the basic information about each lobbying register entry from the overview table. This is necessary here, because the download of the lobbying register entry page needs the link from the Registerzahl field.
- Store the parsed data as JSON file.
- Download all lobbying register entry pages (html) with the unique id as postfix.

**Parse html**

Here the description of the project gets added to the data.

- Open the overview page.
- Parse out the basic information about each lobbying register entry from the overview table. This is necessary here, because the download of the lobbying register entry page needs the link from the Registerzahl field.
- Open all lobbying register entry pages (html).
- Parse out the additional information from all lobbying register pages.
- Store updated data as JSON file.

The lobbying register does not publish the A2 entries, so the scraper never got tested with this and some A2 specific data entries are not part of the parser.

**Download attachments**

Here all attachments linked in the entry pages are stored locally.
- open JSON file.
- Download all attachments.

**Export CSV**

Here the data gets exported as a CSV file.
- Open the data (JSON).
- Save the serialized data as CSV file.

## DATA INPUT

The raw data is from the [Lobbying- und Interessenvertretungs-Register](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/liste!OpenForm&subf=a) of the austrian justice ministry. The data consists of all lobbying activities of
- lobbying companies
- corporations, who employ lobbyists in-house
- self-governing bodies and
- interest groups
since January 1st of 2013. 

More details about the register and it's data can be found [here](http://www.lobbyreg.justiz.gv.at/edikte/ex/edparm3.nsf/h/IR_Hinweise) und [hier](http://www.lobbyreg.justiz.gv.at/edikte/ex/edparm3.nsf/h/ir_Leitfaden/$file/Leitfaden.pdf).

### The Table

The table is the basic data, where most of the data is parsed out. The data is published in the following structure (e. g. first project).

**Example**

| Nr | Bezeichnung/Firma | Registerzahl | Registerabteilung | Details | Letzte Änderung |
|----|-------------------|--------------|-------------------|---------|-----------------|
| 2  | AbbVie GmbH, Wien, (378955m) | [LIVR-00054](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/alldoc/edea53b6f0f5a4b3c1257b3900548209!OpenDocument) | B | Mag. Thomas Haslinger<br/>Mag. Bettina Theresia Kölbl-Resl | 09.11.2015 |

**Attributes**
- Nr: sequential number for each row which is not connected directly with the entry itself.
- Bezeichnung/Firma (Description/Corporation): Name, company or description of company.
- Registerzahl (registry number): the 'Registerzahl' is no unique ID and is taken several times for different organisations, i. e. LIVR-00303 for [Österreichischer Apothekerverband](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/alldoc/2371c20cd6f70fa8c1257bad002ee3a1!OpenDocument) and [Aktienforum - Österreichischer Verband für Aktien-Emittenten und -Investoren](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/alldoc/a61a8fdfd0122b8cc1257bad002edfd6!OpenDocument) lobbying entry.
- Registerabteilung (type of lobbying organisations):
	- A1: Lobbying-Unternehmen bzw. Lobbyisten (Lobbying-corporations or lobbyists)  
	- A2: Aufgabenbereiche der Lobbying-Unternehmen (nicht öffentlich) (areas of activity of lobbying corporations (non-public))  
	- B: Unternehmen bzw. Unternehmens-/(In-House-)Lobbyisten (companies or company-/(in-house-)lobbyists)  
	- C: Selbstverwaltungskörper (self-governing bodies)  
	- D: Interessenverbände (interest groups)  
- Details: name(s) of lobbyist(s) if available.
- Letzte Änderung (last update)

### The lobbying entry pages
Additionally the scraper also takes data from the more detailed pages of each lobbying-register entry, depending on the register type listed at the end.

**Attributes**
- Bekannt gemacht am (announced at): we guess it is the date, when (A1, C)
- Name/Firma (name/company): (A1, B, C)
- Firmenbuchnummer (company register number): only valid austrian company register numbers. (A1, B)
- Firmensitz (register office): place of the register office (A1, C, D)
- Geschäftsanschrift (postal address): address for postal activities. (A1, C, D)
- Beginn des Geschäftsjahres (start of business year): in form of DD.MM. (A1, B)
- Gesetzliche Grundlage (legal foundation): legal foundation for founding of the self-governing body. (C)
- Tätigkeitsbereich (area of activity): short description of professional or business activities and/or the contractual or statuatory area of activity. (A1, B, D)
- Verhaltenskodex (code of conduct): (A1, B)
- Homepage (A1, B, C, D)
- Unternehmenslobbyist/en (corporate lobbyist/s): Lobbyists and/or in-house lobbyists with name (first name familyname) and birthday. One each line. (A1, B)
- Lobbying-Umsatz (lobbying revenue): revenue made through lobbying activities last business year. (A1)
- Anzahl der bearbeiteten Lobbying-Aufträge (number of lobbying orders): number of lobbying orders from last year. (A1) 
- Anzahl Interessenvertreter (number of lobbyists): full number of persons who where mostly lobbying in the last year. (C, D)
- Lobbying-Aufwand > EUR 100.000 (lobbying costs > 100.000€): were the expenses for the last business year were more than 100.000€. (B)
- Kosten der Interessenvertretung (costs of lobbying): from accountant or other statuatory or legal controll body confirmed costs from lobbying. (C, D)
- PDF-Anhänge (pdf attachments): link(s) to submited pdf-attachments. (C)
- Unterorganisationen (sub-organisations): one each line (C, D)
- Kommentar (comment): more detailed information for aggregated data entries. (C, D)

### Soundness
So far, we can not say much about the data quality (completeness, accuracy, etc.), but there are also no reasons to doubt the entries.

The lobbying register does not publish the A2 entries.

**Data errors found**
- Looks like the register number is sometimes wrong, cause it normally should be a unique number. (i. e. LIVR-00303 for [Österreichischer Apothekerverband](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/alldoc/2371c20cd6f70fa8c1257bad002ee3a1!OpenDocument) and [Aktienforum - Österreichischer Verband für Aktien-Emittenten und -Investoren](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/alldoc/a61a8fdfd0122b8cc1257bad002edfd6!OpenDocument))

## DATA OUTPUT

**raw html**

The scraper downloads all raw html of each lobbying register entry and the overview page.

**lobbying data JSON**

The parsed data is stored in an easy-to-read JSON file for further usage.
```
[
	{
		'ID': serial number created from scraper when parsing through the table rows.
		'entryDescription': description of the entry.
		'lobbyingOrgaType': type of lobbying organisation (A1, A2, B, C, D).
		'url': url of the detail page.
		'lastUpdate': last update of detail page.
		'registryNumber': number of company in lobbying-register.
		'orgaName': name of organisation.
		'companyRegisterNumber': number of the national company register.
		'businessActivities': area in which the organisation is active in business.
		'postalAddress': postal address.
		'registeredOffice': Place, where the company is officially registered.
		'businessYearStart': Day, when the business year starts.
		'legalFoundation': 
		'codeOfConduct': Code of conduct.
		'website': url of the website
		'lobbyists': list of lobbyists.
		'lobbyingRevenue': Revenue from lobbying.
		'lobbyingRequests': Number of lobbying requests.
		'numLobbyists': Number of lobbyists.
		'lobbyingCostsGreater100000': 
		'lobbyingCosts': Costs of lobbying.
		'suborganisations': List of suborganisations.
		'attachmentUrls': url to an attachment.
		'comment': comment to precise other fields.
	},
]
```

**lobbying data csv**

The parsed data is stored in a human-readable CSV file for further usage.

columns (see attribute description above):
- ID
- entryDescription
- orgaName
- businessActivities
- lobbyingOrgaType
- lobbyists
- lobbyingRevenue
- lobbyingRequests
- numLobbyists
- lobbyingCostsGreater100000
- lobbyingCosts
- registryNumber
- companyRegisterNumber
- suborganisations
- legalFoundation
- codeOfConduct
- registeredOffice
- website
- postalAddress
- lastUpdate
- dateAnnounced
- businessYearStart
- url
- attachmentUrls
- comment

row: one lobbying register entry each row.

## CONTRIBUTION
In the spirit of free software, everyone is encouraged to help improve this project.

Here are some ways you can contribute:

- by reporting bugs
- by suggesting new features
- by translating to a new language
- by writing or editing documentation
- by analyzing the data
- by visualizing the data
- by writing code (**no pull request is too small**: fix typos in the user interface, add code comments, clean up inconsistent whitespace)
- by refactoring code
- by closing issues
- by reviewing pull requests
- by enriching the data with other data sources

When you are ready, submit a [pull request](https://github.com/OKFNat/lobbyscraper/pulls).

### Submitting an Issue

We use the [GitHub issue tracker](https://github.com/OKFNat/lobbyscraper/issues) to track bugs and features. Before submitting a bug report or feature request, check to make sure it hasn't already been submitted. When submitting a bug report, please try to provide a screenshot that demonstrates the problem. 

## COPYRIGHT
All content is openly licensed under the [Creative Commons Attribution 4.0](http://creativecommons.org/licenses/by/4.0/) license, unless otherwisely stated.

All sourcecode is free software: you can redistribute it and/or modify it under the terms of the MIT License.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Visit [http://opensource.org/licenses/MIT](http://opensource.org/licenses/MIT) to learn more about the MIT License.

## SOURCES

**Gute Taten für gute Daten**
- [Gute Taten für gute Daten](http://okfn.at/gutedaten/)

**Lobbying**
- [Transparency International Austria](https://www.ti-austria.at/): Austrian Chapter of Transparency International.
- [Lobbying in Austria](https://www.ti-austria.at/wp-content/uploads/2016/01/Lobbying-in-Austria.pdf): Publikation von Transparency International Austria

**Documentation**
- [Original Data Source](http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf/liste!OpenForm&subf=a)
- [Hinweise zum Lobbyisten und Interessenvertretungsregister](http://www.lobbyreg.justiz.gv.at/edikte/ex/edparm3.nsf/h/IR_Hinweise)
- [Leitfaden Lobbying- und Interessenvertretungsregister Erfassungsanwendung](http://www.lobbyreg.justiz.gv.at/edikte/ex/edparm3.nsf/h/ir_Leitfaden/$file/Leitfaden.pdf).

## REPOSITORY
- [README.md](README.md): Overview of repository
- [lobbyscraper.py](code/lobbyscraper.py): the scraper
- [CHANGELOG.md](CHANGELOG.md)
- [LICENSE](LICENSE)

## CHANGELOG
See the [whole history](CHANGELOG.md). Next the actual version.

### Version 0.2: 2016-04-26
- update documentation: commenting code, update README.md
- add csv export
- re-factor code
- add attachment download
- add SQLite export





CHANGELOG.md
### HEAD

### Version 0.2: 2016-04-26
- update documentation: commenting code, update README.md
- add csv export
- re-factor code
- add attachment download
- add SQLite export

### Version 0.1: 2015-05-04
**init repo**


code/lobbyscraper.py
#!/bin/env python2
# -*- coding: utf-8 -*-

"""
This is a scraper for the Austrian Lobbying-Register. It fetches the HTML, saves it locally
and converts the relevant data into a json file.
"""

import re
from datetime import datetime, date
import time
import json
import os
from bs4 import BeautifulSoup
import urllib2
import dataset

__author__ = "Stefan Kasberger"
__copyright__ = "Copyright 2015"
__license__ = "MIT"
__version__ = "0.2"
__maintainer__ = "Stefan Kasberger"
__email__ = "mail@stefankasberger.at"
__status__ = "Production" # 'Development', 'Production' or 'Prototype'

###    GLOBAL   ###

ROOT_FOLDER = os.path.dirname(os.getcwd())
FOLDER_RAW_HTML = ROOT_FOLDER + '/data/raw/html/'
FOLDER_RAW_PDF = ROOT_FOLDER + '/data/raw/pdf/'
FOLDER_JSON = ROOT_FOLDER + '/data/json/'
FOLDER_CSV = ROOT_FOLDER + '/data/csv/'
FILENAME_BASE = 'lobbyingregister'
BASE_URL = 'http://www.lobbyreg.justiz.gv.at/edikte/ir/iredi18.nsf'
BASE_URL_ATTACHMENTS = 'http://www.lobbyreg.justiz.gv.at'
QUERY_URL = BASE_URL+'/liste!OpenForm&subf=a'
DELAY_TIME = 2 # in seconds
# TS = datetime.now().strftime('%Y-%m-%d-%H-%M')
TS = '2016-04-19-19-25'

###    FUNCTIONS   ###

def SetupEnvironment():
    """Sets up the folder structure and working environment.
    """
    if not os.path.exists(FOLDER_RAW_HTML):
        os.makedirs(FOLDER_RAW_HTML)
    if not os.path.exists(FOLDER_RAW_PDF):
        os.makedirs(FOLDER_RAW_PDF)
    if not os.path.exists(FOLDER_JSON):
        os.makedirs(FOLDER_JSON)
    if not os.path.exists(FOLDER_CSV):
        os.makedirs(FOLDER_CSV)

def FetchHtml(url):
    """Fetches html url via urllib2.
    
    Args:
        url: url to fetch (string).
    
    Returns:
        html: html string as unicode
    """
    response = urllib2.urlopen(url)
    html = response.read().decode('utf-8')
    time.sleep(DELAY_TIME)
    
    return html

def FetchHtmlOverview(url, folder):
    """Fetches html from the overview list of the lobbyingregister entries and saves it locally.
    
    Args:
        url: url to fetch (string).
        folder: directory to save the html files in.    
    """
    rawHtml = FetchHtml(url)

    if not os.path.exists(folder):
        os.makedirs(folder)
    Save2File(rawHtml, folder+'/overview-page.htm')

def FetchHtmlEntries(entries, folder):
    """Fetches html from an entry in the table and saves it locally with the unique id as postfix.
    
    Args:
        entries: list with sequencial dict() of organisations.
        folder: to save the html in.
    """
    for entry in entries:
        html = FetchHtml(entry['url'])
        Save2File(html, folder+'/entry-'+str(entry['ID'])+'.htm')

def Save2File(data, filename):
    """Saves data on specified place on harddrive.
    
    Args:
        data: string to save.
        filename: string of the filepath.
    """
    text_file = open(filename, "w")
    text_file.write(data.encode('utf-8'))
    text_file.close()

def ReadFile(filename):
    """Reads file and returns the text.
    
    Args:
        filename: name of the file
    
    Returns:
        string: content of file as string
    """
    f = open(filename, 'r')
    string = f.read()

    return string

def ReadEntryFilesInFolder(folder):
    """Reads-in all entry html-files from specified folder.
    
    Args:
        folder: folder where the html-files are stored in.
    
    Returns:
        sortedList: list[] of html texts sorted by file-postfix, which is the unique id of the entry from the table.
    """
    htmlList = []
    sortedList = []

    for filename in os.listdir(folder):
        if filename.find('entry-') >= 0:
            rawHtml = ReadFile(folder+'/'+filename)
            fileIndex = filename.split('.')[0].split('-')[1]
            htmlList.append((int(fileIndex), rawHtml))
    
    # sort list of duppels after first element (the filename postfix) and store to list[]
    htmlList = sorted(htmlList, key=lambda htmlList: htmlList[0])
    for idx, html in htmlList:
        sortedList.append(html)

    return sortedList

def ParseTable(html):
    """Parses the needed facts out of the overview table.
    
    Args:
        html: html (string) to parse. 
    
    Returns:
        list[] of dict() of entries.
            'ID': serial number created from scraper when parsing through the table rows.
            'entryDescription': description of the entry.
            'lobbyingOrgaType': type of lobbying organisation (A1, A2, B, C, D).
            'url': url of the detail page.
            'lastUpdate': last update of detail page.
            'registryNumber': number of company in lobbying-register.
    """
    lobbyList = []
    counter = 1

    soup = BeautifulSoup(html)
    
    # loop over table rows
    for tr in soup.tbody.find_all('tr'):
        tds = tr.find_all('td')

        # assign variables from html table to dict
        entry = {}
        entry['ID'] =  str(counter)
        entry['entryDescription'] =  unicode(tds[1].string)
        entry['lobbyingOrgaType'] =  tds[3].string
        entry['url'] =  BASE_URL+'/'+tds[2].a['href']
        entry['lastUpdate'] = str(datetime.strptime(tds[5].string, '%d.%m.%Y'))
        entry['registryNumber'] = tds[2].string
        
        lobbyList.append(entry)
        counter += 1

    return lobbyList

def ParseEntries(htmlList, entries):
    """Parses the needed facts out of the organisation html.
    
    Args:
        htmlList: list() of html strings.
        entries: list() of dict() of entries.
    
    Returns:
        list[] of dict() of entries.
            'ID': serial number created from scraper when parsing through the table rows.
            'entryDescription': description of the entry.
            'lobbyingOrgaType': type of lobbying organisation (A1, A2, B, C, D).
            'url': url of the detail page.
            'lastUpdate': last update of detail page.
            'registryNumber': number of company in lobbying-register.
            'orgaName': name of organisation.
            'companyRegisterNumber': number of the national company register.
            'businessActivities': area in which the organisation is active in business.
            'postalAddress': postal address.
            'registeredOffice': Place, where the company is officially registered.
            'businessYearStart': Day, when the business year starts.
            'legalFoundation': 
            'codeOfConduct': Code of conduct.
            'website': url of the website
            'lobbyists': list of lobbyists.
            'lobbyingRevenue': Revenue from lobbying.
            'lobbyingRequests': Number of lobbying requests.
            'numLobbyists': Number of lobbyists.
            'lobbyingCostsGreater100000': 
            'lobbyingCosts': Costs of lobbying.
            'suborganisations': List of suborganisations.
            'attachmentUrls': url to an attachment.
            'comment': comment to precise other fields.
    """
    entriesList = []

    for entry in entries:
        soup = BeautifulSoup(htmlList[int(entry['ID'])-1])
        html = unicode(soup)
       
        # regex testing type of registry department: B, C 
        regDepartment = re.findall(r'Registerabteilung:</strong></dt>\n<dd><strong>(.*)</strong></dd></dl>', html)
        if regDepartment:
            if entry['lobbyingOrgaType'] != regDepartment[0]:
                print 'ERROR: lobbying organisation type differs!'

        # regex testing register number: B, C
        regNum = re.findall(r'Registerzahl:</strong></dt>\n<dd><strong>(.*)</strong></dd></dl>', html)
        if regNum:
            if entry['registryNumber'] != regNum[0]:
                print 'ERROR: company register number differs!'

        # regex name: A1, B, C
        name = re.findall(r'Name.*:</strong></dt>\n<dd><strong>(.*)</strong></dd></dl>', html)
        if name:
            entry['orgaName'] = name[0]

        # regex date announced: A1, C
        announceDate = re.findall(r'Bekannt gemacht am:</dt>\n<dd>(.*)</dd>', html)
        if announceDate:
            entry['dateAnnounced'] = str(datetime.strptime(announceDate[0], '%d.%m.%Y'))

        # regex testing last update: A1, B, C
        lastUpdate = re.findall(r'Letzte .*nderung:</dt>\n<dd>(.*)</dd>', html)
        if lastUpdate:
            if entry['lastUpdate'] != str(datetime.strptime(lastUpdate[0], '%d.%m.%Y')):
                print "ERROR: register last update differs!"

        # regex corporate-number: A1, B
        corporateNumber = re.findall(r'Firmenbuchnummer:</dt>\n<dd>(.*)</dd>', html)
        if corporateNumber:
            entry['companyRegisterNumber'] = corporateNumber[0]

        # regex registered office address: A1, C, D
        regOfficeAddress = re.findall(r'itz:</dt>\n<dd>(.*)</dd></dl>', html)
        if regOfficeAddress:
            entry['registeredOffice'] = regOfficeAddress[0]

        # regex mail address: A1, C, D
        postalAddress = re.findall(r'nschrift:</dt>\n<dd>(.*)</dd></dl>', html)
        if postalAddress:
            entry['postalAddress'] = postalAddress[0]

        # regex start business year: A1, B
        startBusinessYear = re.findall(r'ftsjahres:</dt>\n<dd>(.*)</dd></dl>', html)
        if startBusinessYear:
            entry['businessYearStart'] = startBusinessYear[0]

        # regex legal foundation: C
        legalFoundation = re.findall(r'Gesetzliche Grundlage:</dt>\n<dd>(.*)</dd></dl>', html)
        if legalFoundation:
            entry['legalFoundation'] = legalFoundation[0]

        # regex area of activities: A1, B, D
        areaActivities = re.findall(r'bereich:</dt>\n<dd>(.*)</dd></dl>', html)
        if areaActivities:
            entry['businessActivities'] = areaActivities[0]

        # regex codex: A1, B
        codex = re.findall(r'Verhaltenskodex:</dt>\n<dd>(.*)</dd></dl>', html)
        if codex:
            entry['codeOfConduct'] = codex[0]

        # regex website: A1, B, C, D
        website = re.findall(r'Homepage:</dt>\n<dd><a href="(.*)" target="_blank">.*</a></dd></dl>', html)
        if website:
            entry['website'] = website[0]

        # regex lobbyists: A1
        lobbyists = re.findall(r'obbyist.*:</dt>\n<dd>(.*)</dd></dl>', html)
        if lobbyists:
            entry['lobbyists'] = lobbyists[0].split('<br/>')

        # regex lobbying revenue: A1, B
        lobbyingRevenue = re.findall(r'Lobbying-Umsatz:</dt>\n<dd>(.*)</dd></dl>', html)
        if lobbyingRevenue:
            entry['lobbyingRevenue'] = lobbyingRevenue[0]

        # regex lobbying request: A1
        lobbyingRequest = re.findall(r'<dt title="Anzahl der bearbeiteten Lobbying-Auftr.*ge:</dt>\n<dd>(.*)</dd></dl>', html)
        if lobbyingRequest:
            entry['lobbyingRequests'] = lobbyingRequest[0]

        # regex number of lobbyists: C, D
        numLobbyists = re.findall(r'Anzahl Interessenvertreter:</dt>\n<dd>(.*)</dd>', html)
        if numLobbyists:
            entry['numLobbyists'] = numLobbyists[0]

        # regex costs lobbying: B
        costsB = re.findall(r'Lobbying-Aufwand > EUR 100.000:</dt>\n<dd>(.*)</dd>', html)
        if costsB:
            if costsB[0] == 'Ja':
                entry['lobbyingCostsGreater100000'] = True
            if costsB[0] == 'Nein':
                entry['lobbyingCostsGreater100000'] = False
        # regex costs lobbying: C, D
        costsC = re.findall(r'Kosten der Interessenvertretung:</dt>\n<dd>(.*)</dd>', html)
        if costsC:
            entry['lobbyingCosts'] = costsC[0]

        # regex atttachments: C
        attDiv = soup.find(id='UplDL')
        if attDiv:
            entry['attachmentUrlss'] = []
            for link in attDiv.find_all('a'):
                entry['attachmentUrlss'].append(BASE_URL_ATTACHMENTS + link.get('href'))

        # regex subentries: C, D
        subOrganisations = re.findall(r'<dt>Unterorganisation.*</dt>\n<dd>(.*)</dd></dl>', html)
        if subOrganisations:
            entry['suborganisations'] = subOrganisations[0]

        # regex comment: C, D
        comment = re.findall(r'<dt>Kommentar:</dt>\n<dd>(.*)</dd></dl>', html)
        if comment:
            entry['comment'] = comment[0]

        entriesList.append(entry)

    return entriesList

def Save2CSV(entries, filename):
    """Exports the dict into a csv file.
    
    Args:
        entries: list of dict() with all the lobbying register entries.
        filename: name of the file with folder.
    """
    csvString = '"ID","entryDescription","orgaName","businessActivities","lobbyingOrgaType","lobbyists","lobbyingRevenue","lobbyingRequests","numLobbyists","lobbyingCostsGreater100000","lobbyingCosts","registryNumber","companyRegisterNumber","suborganisations","legalFoundation","codeOfConduct","registeredOffice","website","postalAddress","lastUpdate","dateAnnounced","businessYearStart","url","attachmentUrls","comment"\n'

    # iterate over each entry
    for entry in entries:
        uniqueId = '""'
        entryDescription = '""'
        orgaName = '""'
        busActivities = '""'
        lobOrgaType = '""'
        lobbyists = '""'
        lobbyingRevenue = '""'
        lobbyingRequest = '""'
        numLobbyists = '""'
        costsB = '""'
        costsC = '""'
        regNum = '""'
        compRegNumber = '""'
        subOrganisations = '""'
        legalFoundation = '""'
        codex = '""'
        regOfficeAddress = '""'
        lastUpdate = '""'
        website = '""'
        postalAddress = '""'
        lastUpdate = '""'
        announceDate = '""'
        startBusinessYear = '""'
        url = '""'
        attachments = '""'
        comment = '""'

        # read out each attribute
        for elem in entry.keys():
            val = entry[elem]
            if elem == 'ID':
                uniqueId = '"'+val+'"'
            if elem == 'entryDescription':
                entryDescription = '"'+val+'"'
            if elem == 'orgaName':
                orgaName = '"'+val+'"'
            if elem == 'businessActivities':
                busActivities = '"'+val.replace('"', '').replace('\n', '').replace('\r', '')+'"'
            if elem == 'lobbyingOrgaType':
                lobOrgaType = '"'+val+'"'
            if elem == 'lobbyists':
                lobbyists = '"'
                for lob in val:
                    lobbyists += lob+', '
                lobbyists = lobbyists[:-1]
                lobbyists += '"'
            if elem == 'lobbyingRevenue':
                lobbyingRevenue = '"'+val+'"'
            if elem == 'lobbyingRequests':
                lobbyingRequest = '"'+val+'"'
            if elem == 'numLobbyists':
                numLobbyists = '"'+val+'"'
            if elem == 'lobbyingCostsGreater100000':
                costsB = '"'+val+'"'
            if elem == 'lobbyingCosts':
                costsC = '"'+val+'"'
            if elem == 'registryNumber':
                regNum = '"'+val+'"'
            if elem == 'companyRegisterNumber':
                compRegNumber = '"'+val+'"'
            if elem == 'suborganisations':
                subOrganisations = '"'+val+'"'
            if elem == 'legalFoundation':
                legalFoundation = '"'+val+'"'
            if elem == 'codeOfConduct':
                codex = '"'+val.replace('"', '').replace('\n', '').replace('\r', '')+'"'
            if elem == 'registeredOffice':
                regOfficeAddress = '"'+val+'"'
            if elem == 'website':
                website = '"'+val+'"'
            if elem == 'postalAddress':
                postalAddress = '"'+val+'"'
            if elem == 'lastUpdate':
                lastUpdate = '"'+val+'"'
            if elem == 'dateAnnounced':
                announceDate = '"'+val+'"'
            if elem == 'businessYearStart':
                startBusinessYear = '"'+val+'"'
            if elem == 'url':
                url = '"'+val+'"'
            if elem == 'attachmentUrlss':
                attachments = '"'
                for att in val:
                    attachments += att+', '
                attachments +='"'
            if elem == 'comment':
                comment = '"'+val+'"'

        csvString += uniqueId+','+entryDescription+','+orgaName+','+busActivities+','+lobOrgaType+','+lobbyists+','+lobbyingRevenue+','+lobbyingRequest+','+numLobbyists+','+costsB+','+costsC+','+regNum+','+compRegNumber+','+subOrganisations+','+legalFoundation+','+codex+','+regOfficeAddress+','+website+','+postalAddress+','+lastUpdate+','+announceDate+','+startBusinessYear+','+url+','+attachments+','+comment+'\n'

    Save2File(csvString, filename)
    print 'Lobbying data exported as CSV:',filename

def FetchAttachments(entries, folder):
    """Fetches all attachments from the lobbying-register entries.
    
    Args:
        entries: list[] of dict() with all the lobbying-register data.
        folder: directory, where the files are stored in.
    """
    for entry in entries:
        if 'attachmentUrls' in entry.keys():
            for url in entry['attachmentUrls']:
                DownloadFile(url, folder+'/attachment-'+entry['ID']+'_'+url.split('/')[-1])

def DownloadFile(url, filename):
    """Downloads and stores an attachment.
    
    Args:
        url: url of the attachment (string).
        filename: full filepath for the saving.
    """
    if not os.path.exists(os.path.dirname(os.path.abspath(filename))):
        os.makedirs(os.path.dirname(os.path.abspath(filename)))
    response = urllib2.urlopen(url.encode('utf-8'))
    file = open(filename, 'w')
    file.write(response.read())
    file.close()
    time.sleep(DELAY_TIME)

def Save2SQLite(lobbyEntries):
    """Saves the lobbing register entries in a SQLite database. This is not working, because of key-value issues of the dicts().
    
    Args:
        lobbyEntries: list[] of dicts() with lobbying register entries.
    """
    db = dataset.connect('sqlite:///:memory:')
    table = db['lobbyRegisterAT']
    for entry in lobbyEntries:
        print entry
        table.insert(entry)

    # Wien = table.find_one(registeredOffice='Wien')
    # print Wien

###    MAIN   ###

if __name__ == '__main__':
    
    # setup
    startTime = datetime.now()
    print 'start:', startTime
    SetupEnvironment()
    DOWNLOAD_FILES = False
    PARSE_FILES = False
    DOWNLOAD_ATTACHMENTS = False
    EXPORT_DATA = False
 
    if DOWNLOAD_FILES:
        FetchHtmlOverview(QUERY_URL, FOLDER_RAW_HTML+TS) 
        htmlOverview = ReadFile(FOLDER_RAW_HTML+TS+'/overview-page.htm')
        lobbyList = ParseTable(htmlOverview) 
        Save2File(json.dumps(lobbyList, indent=2, ensure_ascii=True), FOLDER_JSON+TS+'_'+FILENAME_BASE+'.json')
        FetchHtmlEntries(lobbyList, FOLDER_RAW_HTML+TS) 
 
    if PARSE_FILES:
        htmlOverview = ReadFile(FOLDER_RAW_HTML+TS+'/overview-page.htm')
        lobbyList = ParseTable(htmlOverview) 
        htmlEntries = ReadEntryFilesInFolder(FOLDER_RAW_HTML+TS) 
        lobbyEntries = ParseEntries(htmlEntries, lobbyList)
        Save2File(json.dumps(lobbyEntries, indent=2, ensure_ascii=True), FOLDER_JSON+TS+'_'+FILENAME_BASE+'.json')
    
    if DOWNLOAD_ATTACHMENTS:    
        lobbyEntries = ReadFile(FOLDER_JSON+TS+'_'+FILENAME_BASE+'.json')
        FetchAttachments(lobbyEntries, FOLDER_RAW_PDF+TS)
 
    if EXPORT_DATA:
        lobbyEntries = json.loads(ReadFile(FOLDER_JSON+TS+'_'+FILENAME_BASE+'.json'))
        Save2CSV(lobbyEntries, FOLDER_CSV+TS+'_'+FILENAME_BASE+'.csv')
        # Save2SQLite(lobbyEntries) # does not run!

    print 'runtime:', (datetime.now() - startTime)

